### DISCUSSION TITLE
How to achieve full duplex with esp-mbedtls?

### DISCUSSION THREAD

**Discussion by yanshay** [esp-rs/esp-mbedtls#49] [Q&A] [answered]
I'm developing an async app using `embassy` and `esp-mbedtls`.
I need to do **full duplex** communication over tls. So receive at the same time as I send.
However, rust wise, the session need to be borrowed mutably for both read and write which prevents calling them simultaniously (even with async which isn't really multithreaded).  So these can't take place at the same time.
With socket there is a reader and a writer and they can be user simultaneously but I couldn't find the same with `esp-mbedtls`.

What I eventually did is write an `embassy task` that handles all communication and communicate with it over channels.
Inside the task it reads from the session and at the same time awaits write requests arriving over the channel (using `select`).
When write requests arrive it stops the read, does the write and then goes back to reading.
When data arrives from the session it puts it on a receive channel to be used by other tasks.

1. Is this a proper way to do that? Is there a risk of losing data if stopping the receive before it's done and switching to send and then back?
2. Is there a more elegant way to do that?

**Comment by AnthonyGrondin** [comment, marked as answer]
With the current `esp-mbedtls` implementation, I think this is the best way to do this for now.

The ideal way, would be to implement something like [TcpSplit](https://github.com/ivmarkov/edge-net/blob/5556e138e00f3f3218a298f54451396d76eb2f1c/edge-nal/src/stack/tcp.rs#L9-L20) or simply a `split()` method on `AsyncConnectedSession`, which would split the connection into a `TlsReader` and `TlsWriter`.

**Reply by yanshay** [reply]
Thanks

**Comment by yanshay** [comment]
Related to this, I'm using `picoserve` as web server for some small application.
If I understand correctly, `picoserve` needs `Reader` and `Writer` to work, meaning, it's impossible to use `esp-mbedtls` to handle TLS, is my understanding correct?
